staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -61,29 +61,88 @@ func NewLRUExpireCacheWithClock(maxSize int, clock Clock) *LRUExpireCache {
 type cacheEntry struct {
 	value      interface{}
 	expireTime time.Time
+
+	// If nil, a computation is in progress. You must wait and then re-get
+	// the cache entry.
+	wait <-chan struct{}
 }
 
 // Add adds the value to the cache at key with the specified maximum duration.
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl), nil})
 }
 
 // Get returns the value at the specified key from the cache if it exists and is not
 // expired, or returns false.
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
+	return c.getCounted(key, 0)
+}
+
+func (c *LRUExpireCache) getCounted(key interface{}, count int) (interface{}, bool) {
+	if count >= 2 {
+		return nil, false
+	}
+	ce, _ := c.getOrBegin(key, 0)
+	if ce == nil {
+		return nil, false
+	}
+	return c.finishGet(key, ce, count)
+}
+
+func (c *LRUExpireCache) finishGet(key interface{}, ce *cacheEntry, count int) (interface{}, bool) {
+	if ce.wait == nil {
+		return ce.value, true
+	}
+	select {
+	case <-ce.wait:
+		return c.getCounted(key, count+1)
```

Comment:
I wonder if there is a way to write this without recursion.

Reply:
WIP means "no need to review yet"; I changed the approach completely like I said...